### DISCUSSION TITLE
Is there a way to convert a parse tree back into the original code?

### DISCUSSION THREAD

**Discussion by cole-kitroser** [tree-sitter/tree-sitter#1651] [Q&A] [answered]
I've been looking through the documentation and what I'm trying to do is to read the code into the parse tree via tree-sitter and then print it back out as a string as the original code. However, as I continue to look through, I am brick walled time and time again, as I see that the parse tree gives us the node's type, but I am unsure about how to get the node's value.

**Comment by ahelwer** [comment, marked as answer]
Unless you have a 1:1 correspondence between your node types and valid syntax in the language (i.e. your language is composed only of keywords) there is no way to do this *only* using the parse tree. The node values (as in, the text that the parse node refers to) are not stored in the tree. The tree only stores start & end points that refer to the text.

**Reply by cole-kitroser** [reply]
Had a feeling this was the answer, but I wanted to double check that I didn't overlook anything. Thanks for answering.
